src/evict/evict_lru.c

```diff
@@ -892,52 +915,55 @@ __evict_clear_walk(WT_SESSION_IMPL *session, bool clear_pos)
+        else {
             /*
-             * If we're at an internal page, then we've just finished all its leafs, so get the
-             * position of the very beginning or the very end of it depending on the direction of
-             * walk.
+             * Remember the last position before clearing it so that we can restart from about the
+             * same point later. evict_saved_ref_check is used as an opaque page id to compare with
+             * it upon restoration for the purpose of stats.
              */
-            if (btree->evict_start_type == WT_EVICT_WALK_NEXT ||
-              btree->evict_start_type == WT_EVICT_WALK_RAND_NEXT) {
-                pos = WT_NPOS_RIGHT;
-                where = "RIGHT";
+            btree->evict_saved_ref_check = (uint64_t)ref;
+
+            if (F_ISSET(ref, WT_REF_FLAG_LEAF)) {
+                /* If we're at a leaf page, use the middle of the page. */
+                pos = WT_NPOS_MID;
+                where = "MIDDLE";
             } else {
-                pos = WT_NPOS_LEFT;
-                where = "LEFT";
+                /*
+                 * If we're at an internal page, then we've just finished all its leafs, so get the
+                 * position of the very beginning or the very end of it depending on the direction
+                 * of walk.
+                 */
+                if (btree->evict_start_type == WT_EVICT_WALK_NEXT ||
+                  btree->evict_start_type == WT_EVICT_WALK_RAND_NEXT) {
+                    pos = WT_NPOS_RIGHT;
+                    where = "RIGHT";
+                } else {
+                    pos = WT_NPOS_LEFT;
+                    where = "LEFT";
+                }
+            }
+            if (!WT_VERBOSE_LEVEL_ISSET(session, WT_VERB_EVICTION, WT_VERBOSE_DEBUG_1))
+                btree->evict_pos = __wt_page_npos(session, ref, pos, NULL, NULL, 0);
+            else {
+                btree->evict_pos =
+                  __wt_page_npos(session, ref, pos, path_str, &path_str_offset, PATH_STR_MAX);
+                __wt_verbose_debug1(session, WT_VERB_EVICTION,
+                  "Evict walk point memorized at position %lf %s of %s page %s ref %p",
+                  btree->evict_pos, where, F_ISSET(ref, WT_REF_FLAG_INTERNAL) ? "INTERNAL" : "LEAF",
+                  path_str, (void *)ref);
             }
-        }
-        if (!WT_VERBOSE_LEVEL_ISSET(session, WT_VERB_EVICTION, WT_VERBOSE_DEBUG_1))
-            btree->evict_pos = __wt_page_npos(session, ref, pos, NULL, NULL, 0);
-        else {
-            btree->evict_pos =
-              __wt_page_npos(session, ref, pos, path_str, &path_str_offset, PATH_STR_MAX);
-            __wt_verbose_debug1(session, WT_VERB_EVICTION,
-              "Evict walk point memorized at position %lf %s of %s page %s ref %p",
-              btree->evict_pos, where, F_ISSET(ref, WT_REF_FLAG_INTERNAL) ? "INTERNAL" : "LEAF",
-              path_str, (void *)ref);
         }
     }
 
```

Comment:
We do page release even for the soft pointer case? Or here the page release should be wrapped in an "else" block?